### ISSUE TITLE
[Bug]: @babel/preset-react inserts jsx-runtime after existing comments, which means they can be lost when bundling

### ISSUE THREAD

**Issue by asos-tomp** [babel/babel#15156]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

@rollup/plugin-babel

### Input code

See [REPL](https://babel.dev/repl#?browsers=Chrome%20106&build=&builtIns=false&corejs=3.21&spec=false&loose=true&code_lz=JYWwDg9gTgLgBAIigUwIYGMYINwChcD0AVHDMgM7zoQgjIB28RBu19lcAYhBHALxwAFAEp-APjgAeACbAAbmMkFZCvLmQAPSLDjTkAM1QBXADbxuEbEA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=react&prettier=false&targets=&version=7.20.3&externalPlugins=&assumptions=%7B%7D) (must manually select "automatic" react runtime)

```jsx
import "react";

/* test comment */
const Foo = () => <div></div>;

export default Foo;
```

### Configuration file name

babel.config.json

### Configuration

```json
{
  "presets": [
    ["@babel/preset-react", { "runtime": "automatic" }]
  ]
}
```

### Current and expected behavior

Expected:
```js
import "react";
import { jsx as _jsx } from "react/jsx-runtime";

/* test comment */
const Foo = () => /*#__PURE__*/_jsx("div", {});
export default Foo;
```
Received:
```js
import "react";

/* test comment */
import { jsx as _jsx } from "react/jsx-runtime";
const Foo = () => /*#__PURE__*/_jsx("div", {});
export default Foo;
```

I would expect the input comment to remain above the `Foo` function, rather than have the `react/jsx-runtime` import interleaved.   

The changed order causes `rollup` to remove the comment, for sanctity of tree-shaking.  Please see [issue raised with rollup](https://github.com/rollup/rollup/issues/4703) for context.

N.B. When `preset-env` is enabled, the imports/comments _are_ in the correct order:
```js
"use strict";

exports.__esModule = true;
exports.default = void 0;
require("react");
var _jsxRuntime = require("react/jsx-runtime");
/* test comment */
const Foo = () => /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {});
var _default = Foo;
exports.default = _default;
```

### Environment

@babel/core: 7.14.6
@babel/preset-react: 7.18.6
MacOS Monterey
Node 18.x

### Possible solution

...will see what I can do...

### Additional context

Perhaps needs a fixture added [here](https://github.com/babel/babel/blob/main/packages/babel-plugin-transform-react-jsx/test/fixtures/react-automatic/.should-properly-handle-comments-adjacent-to-children/input.js)?

**Comment by JLHwung** [comment]
This issue is related to [comment attachment](https://github.com/babel/babel/blob/main/packages/babel-parser/ast/comment-attachment.md), note that the printing behaviour is not changed since Babel 7.

I will first explain what happens here.

```js
import "react";

/* test comment */
const Foo = () => <div></div>;

export default Foo;
```

In this example, the `test comment` is attached as both the  `import "react"` ImportDeclaration's trailing comment and the `const Foo = ...` VariableDeclaration's leading comment (see also [AST Explorer](https://astexplorer.net/#/gist/7361921992414a0ee70d1bf859b6c391/41ccfa4271cc665fffd9ff97df1a3292d8f53353) here).

Now when `@babel/generator` prints the AST, it prints ImportDeclaration before VariableDeclaration, so the trailing comment is printed first. Now when we print the VariableDeclaration `const Foo = ...`, its leading comment is not printed because it has been printed before.

> N.B. When preset-env is enabled, the imports/comments are in the correct order:

When `preset-env` is enabled, the commonjs transform replaces `ImportDeclaration` to a new AST node `const react = ...`. I guess the trailing comment of ImportDeclaration is not inherited (which it should), luckily the generator picks up the comment when it prints `const Foo = ...`, so now the comment prints right before `const Foo`.

Now you may think maybe we can prioritize leading comments over trailing comments. Well let's take a look at another example:

```js
import react from "React" // eslint-disable-line import/no-extraneous-dependencies

const Foo = () => <div></div>;
```

We will agree that the `eslint-disable-line` comment should be always printed after `import react`. So Babel should output

```js
import react from "React" // eslint-disable-line import/no-extraneous-dependencies
import { jsx as _jsx } from "react/jsx-runtime";

const Foo = () => /*#__PURE__*/_jsx("div", {});
```
However, from Babel's perspective, this example is no different than yours. As human we use whitespaces to convey our preferences about how the comment should be attached to its adjacent AST nodes, however this is not captured in the AST therefore the generator can't work on both cases.

As a workaround, you can manually add something between the import's and `const Foo`:

```js
import react from "React";

{}
/* test comment */
const Foo = () => <div></div>;
```

the minifier will happily strip `{}` for you.

**Comment by asos-tomp** [comment]
Thank you so much for this in-depth answer; it certainly makes a lot of sense that the parser cannot prefer leading over trailing affinity for comments; at least not without some inference that "end of line" comments have affinity to the line they trail, and that in all other (especially multi-line?) cases comments generally have affinity to something that follows. 

That may be a more common convention, but certainly nothing I'm aware of in any specification.

Perhaps, if anything, it highlights that the `commonjs` transform should be consistent, at least - but this would not resolve my (and perhaps `@rollup/plugin-babel`'s) issue.

Thanks again for your input, I have avenues to explore, at least!   If a PR that explores the option of making a biased choice would be welcome, I would be happy to contribute; but it sounds like it wouldn't be based on any particular rigour.

**Comment by liuxingbaoyu** [comment]
We actually append comments to both the front and rear nodes, and the printer will print them only once. This allows more comments to be left after complex transformations.

> and that in all other (especially multi-line?) cases comments generally have affinity to something that follows.

Yes it's theoretically possible, but I think it's hard to implement. We cannot establish a two-way reference to a comment to detect whether any subsequent nodes hold the comment.

**Comment by asos-tomp** [comment]
Makes sense, thanks.

I guess in this _specific_ use case (injection of `import { jsx } from 'react/jsx-runtime';` via `@babel/preset-react`) it could be said that since the import should not rely on side-effects or have a need for ordering, an inference _could_ be made to inject this as if at the beginning of the module, rather than immediately proceeding the use of `jsx`.   This might "work", and solve the issue at hand, if somewhat subverting the "just in time" ordering of imports.

Rollup evidently _does_ make a strong inference that a comment immediately proceeding an import statement is connected, and thus compiles it out along with the import - as per c[omment by @lukastaegert](https://github.com/rollup/rollup/issues/4703#issuecomment-1306093574) on the corresponding [rollup issue](https://github.com/rollup/rollup/issues/4703) I raised:
> As import statements are ALWAYS removed, the comment is removed with the statement

**Comment by liuxingbaoyu** [comment]
![image](https://user-images.githubusercontent.com/30521560/200508050-17588758-d21f-4c4c-99ed-50ca33b7625d.png)
Actually the import is being added to the beginning of the code, while the import in the last input code holds a reference to the comment.
A crude idea, maybe do a simple preprocessing of the ast in `@babel/core`, but I'm afraid this will affect `// eslint-disable-line xxxx`

**Comment by asos-tomp** [comment]
I guess by "beginning of the code" I would have assumed:
```js
import { jsx as _jsx } from "react/jsx-runtime";
import "react";

/* test comment */
const Foo () => /*#__PURE__*/_jsx("div", {});
export default Foo;
```
...could perhaps be achieved?

**Comment by liuxingbaoyu** [comment]
```
import { jsx as _jsx } from "react/jsx-runtime";
// This file xxxxx
import a
```

This seems to be worse.
I'm not sure the import order would make sense in some edge cases, while the babel addition at the end is more readable for me.

**Comment by asos-tomp** [comment]
Agreed that it would not be wise, in general, due to the ordering/dependency issues that you describe.

Just aware that there's a very specific babel plugin for injecting this `react/jsx-runtime` that could be "special cased" if it's understood it should never be dependent on side-effects created prior the same source file.

**Comment by asos-tomp** [comment]
Perhaps related? https://github.com/babel/babel/issues/5512

**Comment by liuxingbaoyu** [comment]
Yes, it looks like that's a similar issue.
I've tried a few things, it's a bit cumbersome, and only works in a few cases. 😕
At the same time it introduces a lot of complexity and possible unstable output, and I'm not sure it's worth it compared to stating it in the documentation.

**Comment by liuxingbaoyu** [comment]
![image](https://user-images.githubusercontent.com/30521560/203703214-3ca44676-ea2f-4ca7-a9d2-b99dd9ed6371.png)

Yes, this is really annoying, and I didn't even think of a good way.
Currently our comment blank line depends on node, when it is used as trailing Comments, it can print newline normally. But when it is used as leading Comments, empty lines will not be printed, which is obviously worse.
